### PR TITLE
Better input ordering

### PR DESCRIPTION
**Pull request by DaughterOfMars**
# Description of change

Previously, we sorted the inputs when the builder is finished (and in the mana requirement) but this was unnecessary  as we can sort them when selecting instead and just maintain the sorted order.

UPDATE: After some re-evaluation, the above approach is not possible because inputs may be added that have requirements that have not yet been selected, but will be later. This would then need basically the same re-sorting logic. Instead, I implemented a custom collection which holds useful information for ordering the inputs and handles the actual order in the iterator implementations. This does require some additional memory overhead per reference iterator call (but not for consuming iteration): An additional `VecDeque` and `BTreeMap` are allocated, plus one `VecDeque` per non-ed25519 address.

### CODE REVIEW

File: sdk/src/client/api/block_builder/transaction_builder/mod.rs [hunk 1]
```diff
@@ -680,38 +672,38 @@ impl TransactionBuilder {
...
-        let (mut sorted_inputs, account_nft_address_inputs): (Vec<InputSigningData>, Vec<InputSigningData>) =
-            inputs.into_iter().partition(|input_signing_data| {
+    fn insert_input_sorted(&mut self, input: InputSigningData) {
+        let input_required_address = input
+            .output
+            .required_address(
+                self.latest_slot_commitment_id.slot_index(),
+                self.protocol_parameters.committable_age_range(),
+            )
+            // PANIC: safe to unwrap as non basic/account/foundry/nft outputs are already filtered out.
```
Comment (Thoralf-M) [inline sdk/src/client/api/block_builder/transaction_builder/mod.rs]: Regarding the comment, these outputs aren't filtered out here? Same in L 703

Comment (DaughterOfMars) [reply sdk/src/client/api/block_builder/transaction_builder/mod.rs]: good point

Comment (DaughterOfMars) [reply sdk/src/client/api/block_builder/transaction_builder/mod.rs]: though they were never correct, so this isn't any different

Comment (Thoralf-M) [reply sdk/src/client/api/block_builder/transaction_builder/mod.rs]: They were correct when we had the treasury output, but not updated when we removed it
```Rust
    pub fn unlock_conditions(&self) -> Option<&UnlockConditions> {
        match self {
            Self::Basic(output) => Some(output.unlock_conditions()),
            Self::Account(output) => Some(output.unlock_conditions()),
            Self::Anchor(output) => Some(output.unlock_conditions()),
            Self::Foundry(output) => Some(output.unlock_conditions()),
            Self::Nft(output) => Some(output.unlock_conditions()),
            Self::Delegation(output) => Some(output.unlock_conditions()),
        }
    }
```
Should just remove the option here I guess

Comment (DaughterOfMars) [reply sdk/src/client/api/block_builder/transaction_builder/mod.rs]: actually, they are filtered out of the available outputs, so this is true

Comment (Thoralf-M) [reply sdk/src/client/api/block_builder/transaction_builder/mod.rs]: Delegation outputs can't be there?

Comment (DaughterOfMars) [reply sdk/src/client/api/block_builder/transaction_builder/mod.rs]: No, but they are safe to unwrap at least

**Review by Thoralf-M — APPROVED** [review]

**Review by Thoralf-M — CHANGES_REQUESTED** [review]
Failing test

File: sdk/src/client/api/block_builder/transaction_builder/mod.rs [hunk 2]
```diff
@@ -680,38 +672,39 @@ impl TransactionBuilder {
...
+                self.protocol_parameters.committable_age_range(),
+            )
+            // PANIC: safe to unwrap as non basic/account/foundry/nft/delegation outputs are already filtered out.
+            .unwrap()
+            .expect("expiration unlockable outputs already filtered out");
+        if input_required_address.is_ed25519() {
+            if let Some(position) = self
+                .selected_inputs
+                .iter()
+                .position(|i| i.output_id() > input.output_id())
```
Comment (Thoralf-M) [inline sdk/src/client/api/block_builder/transaction_builder/mod.rs]: Why is it comparing the output ids like this?

Comment (DaughterOfMars) [reply sdk/src/client/api/block_builder/transaction_builder/mod.rs]: Yaknow, I don't know 😆 I think I got confused

File: sdk/src/client/api/block_builder/transaction_builder/mod.rs [hunk 3]
```diff
@@ -688,99 +690,117 @@ impl TransactionBuilder {
...
+    pub(crate) fn iter(&self) -> OrderedInputsIter<&InputSigningData> {
+        self.into_iter()
+    }
 
-        for input in account_nft_address_inputs {
-            let required_address = input
-                .output
-                .required_address(commitment_slot_index, committable_age_range)?
-                .expect("expiration unlockable outputs already filtered out");
+    pub(crate) fn insert(&mut self, input: InputSigningData, required_address: Address) {
```
Comment (Thoralf-M) [inline sdk/src/client/api/block_builder/transaction_builder/mod.rs]: Accept the required_address as first param since it's the key?

Comment (DaughterOfMars) [reply sdk/src/client/api/block_builder/transaction_builder/mod.rs]: sure

File: sdk/src/client/api/block_builder/transaction_builder/mod.rs [hunk 4]
```diff
@@ -688,99 +690,117 @@ impl TransactionBuilder {
...
+        if required_address.is_ed25519() {
```
Comment (Thoralf-M) [inline sdk/src/client/api/block_builder/transaction_builder/mod.rs]: I think this need `is_ed25519_backed()` or `if let Some(ed25519) = required_address.backing_ed25519(){}` as they require the same Ed25519 signature

Comment (DaughterOfMars) [reply sdk/src/client/api/block_builder/transaction_builder/mod.rs]: okay

File: sdk/src/client/api/block_builder/transaction_builder/mod.rs [hunk 5]
```diff
@@ -688,99 +690,113 @@ impl TransactionBuilder {
...
+                    for input in self
+                        .other
+                        .remove(&Address::Account(AccountAddress::new(
+                            account_output.account_id_non_null(input.borrow().output_id()),
+                        )))
+                        .into_iter()
+                        .flatten()
+                        .rev()
+                    {
+                        self.queue.push_front(input);
```
Comment (Thoralf-M) [inline sdk/src/client/api/block_builder/transaction_builder/mod.rs:765]: How does this handle multiple account outputs owned by other account outputs?
Like there is 
Account A controlled by an Ed25519 address
Account B controlled by Account A
Account C controlled by Account B
where does it make sure that Account B is before Account C, so the unlock can reference it with a lower index?

Comment (DaughterOfMars) [reply sdk/src/client/api/block_builder/transaction_builder/mod.rs:765]: it's the purpose of the queue. it will get the next input, then find the inputs that depend on it and add those to the front of the queue, so it looks like this:
```
Account A controlled by an Ed25519 address
Account B controlled by Account A
Account C controlled by Account B
Account D controlled by an Ed25519 address
Account E controlled by an Ed25519 address
Account F controlled by Account D

Queue at each next call:
[A, D, E] -> A
[B, D, E] -> B
[C, D, E] -> C
[D, E] -> D
[F, E] -> F
[E] -> E

Comment (Thoralf-M) [reply sdk/src/client/api/block_builder/transaction_builder/mod.rs:765]: But what if the queue is not already ordered as it needs to be (or is that done somewhere)? So like this
Account A controlled by an Ed25519 address
Account C controlled by Account B
Account B controlled by Account A

Comment (DaughterOfMars) [reply sdk/src/client/api/block_builder/transaction_builder/mod.rs:765]: The queue initially only contains the ed25519 controlled outputs, so that would look like this:

```
[A] -> A
[B] -> B
[C] -> C
```

As each dependent is looked up from the map on each iteration

Comment (Thoralf-M) [reply sdk/src/client/api/block_builder/transaction_builder/mod.rs:765]: Ok I added debug logs to understand it better like this
```Rust
    fn next(&mut self) -> Option<Self::Item> {
        log::debug!("queue: {:?}", self.queue);
        if let Some(input) = self.queue.pop_front() {
            // Add associated inputs to the front of the queue
            match &input.borrow().output {
                Output::Account(account_output) => {
                    for input in self
                        .other
                        .remove(&Address::Account(AccountAddress::new(
                            account_output.account_id_non_null(input.borrow().output_id()),
                        )))
                        .into_iter()
                        .flatten()
                        .rev()
                    {
                        self.queue.push_front(input);
                        log::debug!("queue 2: {:?}", self.queue);
```
and this is the output (I removed the actual self.queue here to save space)
```
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG Selecting input OutputId { id: "0x8d144dd04605a0d249abed652c320b580f96c7119d1296bb600eadb1cd1160155a0000000000", transaction_id: TransactionId { id: "0x8d144dd04605a0d249abed652c320b580f96c7119d1296bb600eadb1cd1160155a000000", slot_index: SlotIndex(90) }, output_index: 0 }
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::transition DEBUG No transition of OutputId { id: "0x8d144dd04605a0d249abed652c320b580f96c7119d1296bb600eadb1cd1160155a0000000000", transaction_id: TransactionId { id: "0x8d144dd04605a0d249abed652c320b580f96c7119d1296bb600eadb1cd1160155a000000", slot_index: SlotIndex(90) }, output_index: 0 }/AccountId(0x2222222222222222222222222222222222222222222222222222222222222222) as output already exists
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG Selecting input OutputId { id: "0x1f0cdc1bf552449906ea7139f6a29ddd0e7619915841b6de8cf73032226df55f5a0000000000", transaction_id: TransactionId { id: "0x1f0cdc1bf552449906ea7139f6a29ddd0e7619915841b6de8cf73032226df55f5a000000", slot_index: SlotIndex(90) }, output_index: 0 }
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::transition DEBUG No transition of OutputId { id: "0x1f0cdc1bf552449906ea7139f6a29ddd0e7619915841b6de8cf73032226df55f5a0000000000", transaction_id: TransactionId { id: "0x1f0cdc1bf552449906ea7139f6a29ddd0e7619915841b6de8cf73032226df55f5a000000", slot_index: SlotIndex(90) }, output_index: 0 }/AccountId(0x1111111111111111111111111111111111111111111111111111111111111111) as output already exists
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG Adding Account(AccountId(0x2222222222222222222222222222222222222222222222222222222222222222)) from input OutputId { id: "0x1f0cdc1bf552449906ea7139f6a29ddd0e7619915841b6de8cf73032226df55f5a0000000000", transaction_id: TransactionId { id: "0x1f0cdc1bf552449906ea7139f6a29ddd0e7619915841b6de8cf73032226df55f5a000000", slot_index: SlotIndex(90) }, output_index: 0 }
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG Selecting input OutputId { id: "0xe21a070f8a2d685ea29a701f171271cce9823f3d83e00ccc9327e96ec7e63eb05a0000000000", transaction_id: TransactionId { id: "0xe21a070f8a2d685ea29a701f171271cce9823f3d83e00ccc9327e96ec7e63eb05a000000", slot_index: SlotIndex(90) }, output_index: 0 }
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::transition DEBUG No transition of OutputId { id: "0xe21a070f8a2d685ea29a701f171271cce9823f3d83e00ccc9327e96ec7e63eb05a0000000000", transaction_id: TransactionId { id: "0xe21a070f8a2d685ea29a701f171271cce9823f3d83e00ccc9327e96ec7e63eb05a000000", slot_index: SlotIndex(90) }, output_index: 0 }/AccountId(0x3333333333333333333333333333333333333333333333333333333333333333) as output already exists
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG Adding Account(AccountId(0x1111111111111111111111111111111111111111111111111111111111111111)) from input OutputId { id: "0xe21a070f8a2d685ea29a701f171271cce9823f3d83e00ccc9327e96ec7e63eb05a0000000000", transaction_id: TransactionId { id: "0xe21a070f8a2d685ea29a701f171271cce9823f3d83e00ccc9327e96ec7e63eb05a000000", slot_index: SlotIndex(90) }, output_index: 0 }
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::requirement DEBUG Adding Account(AccountId(0x3333333333333333333333333333333333333333333333333333333333333333)) from output
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::requirement DEBUG Adding Account(AccountId(0x1111111111111111111111111111111111111111111111111111111111111111)) from output
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::requirement DEBUG Adding Account(AccountId(0x2222222222222222222222222222222222222222222222222222222222222222)) from output
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::requirement DEBUG Fulfilling requirement Account(AccountId(0x2222222222222222222222222222222222222222222222222222222222222222))
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::requirement::account DEBUG AccountId(0x2222222222222222222222222222222222222222222222222222222222222222) requirement already fulfilled by OutputId { id: "0x8d144dd04605a0d249abed652c320b580f96c7119d1296bb600eadb1cd1160155a0000000000", transaction_id: TransactionId { id: "0x8d144dd04605a0d249abed652c320b580f96c7119d1296bb600eadb1cd1160155a000000", slot_index: SlotIndex(90) }, output_index: 0 }
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::requirement DEBUG Fulfilling requirement Account(AccountId(0x1111111111111111111111111111111111111111111111111111111111111111))
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::requirement::account DEBUG AccountId(0x1111111111111111111111111111111111111111111111111111111111111111) requirement already fulfilled by OutputId { id: "0x1f0cdc1bf552449906ea7139f6a29ddd0e7619915841b6de8cf73032226df55f5a0000000000", transaction_id: TransactionId { id: "0x1f0cdc1bf552449906ea7139f6a29ddd0e7619915841b6de8cf73032226df55f5a000000", slot_index: SlotIndex(90) }, output_index: 0 }
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::requirement DEBUG Fulfilling requirement Account(AccountId(0x3333333333333333333333333333333333333333333333333333333333333333))
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::requirement::account DEBUG AccountId(0x3333333333333333333333333333333333333333333333333333333333333333) requirement already fulfilled by OutputId { id: "0xe21a070f8a2d685ea29a701f171271cce9823f3d83e00ccc9327e96ec7e63eb05a0000000000", transaction_id: TransactionId { id: "0xe21a070f8a2d685ea29a701f171271cce9823f3d83e00ccc9327e96ec7e63eb05a000000", slot_index: SlotIndex(90) }, output_index: 0 }
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::requirement DEBUG Fulfilling requirement Account(AccountId(0x1111111111111111111111111111111111111111111111111111111111111111))
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::requirement::account DEBUG AccountId(0x1111111111111111111111111111111111111111111111111111111111111111) requirement already fulfilled by OutputId { id: "0x1f0cdc1bf552449906ea7139f6a29ddd0e7619915841b6de8cf73032226df55f5a0000000000", transaction_id: TransactionId { id: "0x1f0cdc1bf552449906ea7139f6a29ddd0e7619915841b6de8cf73032226df55f5a000000", slot_index: SlotIndex(90) }, output_index: 0 }
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::requirement DEBUG Fulfilling requirement Account(AccountId(0x2222222222222222222222222222222222222222222222222222222222222222))
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::requirement::account DEBUG AccountId(0x2222222222222222222222222222222222222222222222222222222222222222) requirement already fulfilled by OutputId { id: "0x8d144dd04605a0d249abed652c320b580f96c7119d1296bb600eadb1cd1160155a0000000000", transaction_id: TransactionId { id: "0x8d144dd04605a0d249abed652c320b580f96c7119d1296bb600eadb1cd1160155a000000", slot_index: SlotIndex(90) }, output_index: 0 }
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::requirement DEBUG Fulfilling requirement NativeTokens
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::requirement::native_tokens DEBUG Native tokens requirement already fulfilled
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::requirement DEBUG Fulfilling requirement Amount
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::requirement::amount DEBUG Amount requirement already fulfilled
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::requirement DEBUG Fulfilling requirement Mana
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::requirement::mana DEBUG Mana requirement selected mana: 0, required mana: 0
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::requirement::mana DEBUG Mana requirement already fulfilled
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::remainder DEBUG input_nts: {}
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::remainder DEBUG output_nts: {}
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder::remainder DEBUG No remainder required
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue 2
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:
2024-03-13 19:32:14 (UTC) iota_sdk::client::api::block_builder::transaction_builder DEBUG queue:

```
That's only for 3 inputs, seem like quite a lot of action, isn't that worse than ordering once at the end?

Comment (DaughterOfMars) [reply sdk/src/client/api/block_builder/transaction_builder/mod.rs:765]: You have a point. Perhaps we can iterate unsorted for times when it doesn't matter

Comment (DaughterOfMars) [reply sdk/src/client/api/block_builder/transaction_builder/mod.rs:765]: I separated the sorted and unsorted iteration so now it will only do the sorting for places that need it

File: sdk/src/client/api/block_builder/transaction_builder/mod.rs [hunk 6]
```diff
@@ -688,99 +690,108 @@ impl TransactionBuilder {
...
+                _ => (),
+            };
+            return Some(input);
+        }
+        if let Some(mut entry) = self.other.first_entry() {
+            if let Some(input) = entry.get_mut().pop_front() {
+                if entry.get().is_empty() {
+                    self.other.pop_first();
+                }
+                return Some(input);
```
Comment (Thoralf-M) [inline sdk/src/client/api/block_builder/transaction_builder/mod.rs]: Can you please add some comments to the code what it's doing, bit hard to follow what this is for
Also some general thing like we had before `Inputs need to be sorted before signing, because the reference unlock conditions can only reference a lower index`

Comment (Thoralf-M) [reply sdk/src/client/api/block_builder/transaction_builder/mod.rs]: Thanks

**Review by Thoralf-M — DISMISSED** [review]

**Review by Thoralf-M — APPROVED** [review]

**Review by Alex6323 — APPROVED** [review]
